pkg/virt-controller/watch/replicaset.go

```diff
@@ -0,0 +1,372 @@
+	vmRSInformer cache.SharedIndexInformer
+	recorder     record.EventRecorder
+}
+
+func (c *VMReplicaSet) Run(threadiness int, stopCh chan struct{}) {
+	defer kubecli.HandlePanic()
+	defer c.queue.ShutDown()
+	logging.DefaultLogger().Info().Msg("Starting controller.")
+
+	// Wait for cache sync before we start the pod controller
+	cache.WaitForCacheSync(stopCh, c.vmInformer.HasSynced, c.vmRSInformer.HasSynced)
+
+	// Start the actual work
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(c.runWorker, time.Second, stopCh)
+	}
+
+	<-stopCh
+	logging.DefaultLogger().Info().Msg("Stopping controller.")
+}
+
+func (c *VMReplicaSet) runWorker() {
+	for c.Execute() {
+	}
+}
+
+func (c *VMReplicaSet) Execute() bool {
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(key)
+	if err := c.execute(key.(string)); err != nil {
+		logging.DefaultLogger().Info().Reason(err).Msgf("reenqueuing VirtualMachineReplicaSet %v", key)
+		c.queue.AddRateLimited(key)
+	} else {
+		logging.DefaultLogger().Info().V(4).Msgf("processed VirtualMachineReplicaSet %v", key)
+		c.queue.Forget(key)
+	}
+	return true
+}
+
+func (c *VMReplicaSet) execute(key string) error {
+
+	obj, exists, err := c.vmRSInformer.GetStore().GetByKey(key)
+	if err != nil {
+		return nil
+	}
+	if !exists {
+		// nothing we need to do. It should always be possible to re-create this type of controller
+		return nil
+	}
+	rs := obj.(*virtv1.VirtualMachineReplicaSet)
+
+	//TODO default rs if necessary, the aggregated apiserver will do that in the future
+	if rs.Spec.Template == nil || rs.Spec.Selector == nil || len(rs.Spec.Template.ObjectMeta.Labels) == 0 {
+		logging.DefaultLogger().Object(rs).Error().Msg("Invalid controller spec, will not retry processing it.")
+		return nil
+	}
+
+	selector, err := v1.LabelSelectorAsSelector(rs.Spec.Selector)
+	if err != nil {
+		return nil
+	}
+
+	// get all potentially interesting VMs from the cache
+	vms, err := c.listVMsFromNamespace(rs.ObjectMeta.Namespace)
+
+	if err != nil {
+		return err
+	}
+
+	// make sure we only consider active VMs
+	vms = c.filterActiveVMs(vms)
+
+	// make sure the selector of the controller matches and the VMs match
+	vms = c.filterMatchingVMs(selector, vms)
+
+	// Scale up or down
+	scaleErr := c.scale(rs, vms)
+
+	obj, err = model.Clone(rs)
+	if err != nil {
+		return err
+	}
+	rsCopy := obj.(*virtv1.VirtualMachineReplicaSet)
+
+	return c.updateStatus(rsCopy, vms, scaleErr)
+}
+
+func (c *VMReplicaSet) scale(rs *virtv1.VirtualMachineReplicaSet, vms []virtv1.VM) error {
+
+	diff := c.calcDiff(rs, vms)
+
+	if diff == 0 {
+		return nil
+	}
+
+	// Every delete request can fail, give the channel enough room, to not block the go routines
+	errChan := make(chan error, abs(diff))
+
+	var wg sync.WaitGroup
+	wg.Add(abs(diff))
+
+	if diff > 0 {
+		// We have to delete VMs
+		for i := 0; i < diff; i++ {
+			go func(idx int) {
+				defer wg.Done()
+				deleteCandidate := vms[idx]
```

Comment:
In the future we may want to introduce delete strategies here, things like "delete oldest first"

Reply:
Yep, that, or we can come up with a "best" sorting for that use-case. The k8s replica set does that on the pod level. Did not think about it in detail, though.